File: app/src/main/java/hu/kole/test/java/adapter/SectionAdapter.java
package hu.kole.test.java.adapter;

import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.recyclerview.extensions.ListAdapter;
import android.support.v7.widget.SimpleItemAnimator;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

import hu.kole.nestedlist.BaseNestedListAdapter;
import hu.kole.nestedlist.NestedItemCallback;
import hu.kole.test.R;
import hu.kole.test.data.Item;
import hu.kole.test.data.Section;
import hu.kole.test.databinding.ViewSectionBinding;

public class SectionAdapter extends BaseNestedListAdapter<Section, Item, BaseNestedListAdapter.BaseNestedViewHolder<Section, Item>> {
    public SectionAdapter() {
        super(new SectionDiffCallback());
    }

    @NotNull
    @Override
    public List<Item> getNestedItems(int position) {
        return getItem(position).getData();
    }

    @NonNull
    @Override
    public BaseNestedViewHolder<Section, Item> onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        return new SectionViewHolder((ViewSectionBinding) DataBindingUtil.inflate(LayoutInflater.from(viewGroup.getContext()), R.layout.view_section, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull BaseNestedViewHolder<Section, Item> holder, int position) {
        holder.setup(position, getItem(position));
    }
}

class SectionViewHolder extends BaseNestedListAdapter.BaseNestedViewHolder<Section, Item> {
    private ViewSectionBinding mBinding;
    private ItemAdapter mNestedAdapter;

    public SectionViewHolder(@NotNull ViewSectionBinding binding) {
        super(binding.getRoot());
        mBinding = binding;
    }

    @NotNull
    @Override
    public ListAdapter<Item, ?> getNestedListAdapter() {
        if (mNestedAdapter == null) {
            mNestedAdapter = new ItemAdapter();
            mNestedAdapter.setHasStableIds(true);
        }

        return mNestedAdapter;
    }

    @Override
    public void setup(int position, @Nullable Section section) {
        if (section != null) {
            SimpleItemAnimator animator = (SimpleItemAnimator) mBinding.recyclerView.getItemAnimator();

            if (animator != null) {
                animator.setSupportsChangeAnimations(false);
            }

            mBinding.recyclerView.setHasFixedSize(true);

            mBinding.recyclerView.setAdapter(getNestedListAdapter());
            mBinding.setSection(section);
            mBinding.executePendingBindings();

            this.getNestedListAdapter().submitList(section.getData());
        }
    }
}

class SectionDiffCallback extends NestedItemCallback<Section> {

    @Override
    public boolean areNestedContentsTheSame(Section oldSection, Section newSection) {
        return oldSection.getData().equals(newSection.getData());
    }

    @Override
    public boolean areItemsTheSame(@NonNull Section oldSection, @NonNull Section newSection) {
        return oldSection.equals(newSection);
    }

    @Override
    public boolean areContentsTheSame(@NonNull Section oldSection, @NonNull Section newSection) {
        return oldSection.equals(newSection);
    }
}

File: app/src/main/java/hu/kole/test/adapter/SectionAdapter.kt
package hu.kole.test.adapter

import android.databinding.DataBindingUtil
import android.support.v7.recyclerview.extensions.ListAdapter
import android.support.v7.widget.SimpleItemAnimator
import android.view.LayoutInflater
import android.view.ViewGroup
import hu.kole.nestedlist.BaseNestedListAdapter
import hu.kole.nestedlist.NestedItemCallback
import hu.kole.test.R
import hu.kole.test.data.Item
import hu.kole.test.data.Section
import hu.kole.test.databinding.ViewSectionBinding

class SectionAdapter: BaseNestedListAdapter<Section, Item, BaseNestedListAdapter.BaseNestedViewHolder<Section, Item>>(SectionDiffCallback()) {
    override fun getNestedItems(position: Int): List<Item> {
        return getItem(position).data
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseNestedViewHolder<Section, Item> {
        return SectionViewHolder(DataBindingUtil.inflate(LayoutInflater.from(parent.context), R.layout.view_section, parent, false))
    }

    override fun onBindViewHolder(holder: BaseNestedViewHolder<Section, Item>, position: Int) {
        holder.setup(position, getItem(position))
    }
}

class SectionViewHolder(private val binding: ViewSectionBinding): BaseNestedListAdapter.BaseNestedViewHolder<Section, Item>(binding.root) {
    override val nestedListAdapter: ListAdapter<Item, *> = ItemAdapter().apply { setHasStableIds(true) }

    override fun setup(position: Int, section: Section?) {
        section?.let {
            (binding.recyclerView.itemAnimator as SimpleItemAnimator).supportsChangeAnimations = false
            binding.recyclerView.setHasFixedSize(true)
            binding.recyclerView.adapter = nestedListAdapter
            binding.section = section
            binding.executePendingBindings()

            this.nestedListAdapter.submitList(section.data)
        }
    }
}

class SectionDiffCallback: NestedItemCallback<Section>() {
    override fun areNestedContentsTheSame(old: Section, new: Section): Boolean {
        return old.data == new.data
    }

    override fun areItemsTheSame(old: Section, new: Section): Boolean {
        return old == new
    }

    override fun areContentsTheSame(old: Section, new: Section): Boolean {
        return old == new
    }
}